intake_esm/core.py

```diff
@@ -58,31 +57,171 @@ class esm_datastore(intake.catalog.Catalog):
-        self.preprocess = None
-        self.aggregate = None
+        self._grouped = self.df.groupby(self.aggregation_info['groupby_attrs'])
+        self._keys = list(self._grouped.groups.keys())
+        super(esm_datastore, self).__init__(**kwargs)
+
+    def _get_aggregation_info(self):
+        groupby_attrs = []
+        data_format = None
+        format_column_name = None
+        variable_column_name = None
+        aggregations = []
+        aggregation_dict = {}
+        agg_columns = []
+        path_column_name = self.esmcol_data['assets']['column_name']
+
+        if 'format' in self.esmcol_data['assets']:
+            data_format = self.esmcol_data['assets']['format']
+        else:
+            format_column_name = self.esmcol_data['assets']['format_column_name']
+
+        if 'aggregation_control' in self.esmcol_data:
+            aggregation_dict = {}
+            variable_column_name = self.esmcol_data['aggregation_control']['variable_column_name']
+            groupby_attrs = self.esmcol_data['aggregation_control'].get('groupby_attrs', [])
+            aggregations = self.esmcol_data['aggregation_control'].get('aggregations', [])
+            # Sort aggregations to make sure join_existing is always done before join_new
+            aggregations = sorted(aggregations, key=lambda i: i['type'], reverse=True)
+            for agg in aggregations:
+                key = agg['attribute_name']
+                rest = agg.copy()
+                del rest['attribute_name']
+                aggregation_dict[key] = rest
+            agg_columns = list(aggregation_dict.keys())
+
+        if not groupby_attrs:
+            groupby_attrs = self.df.columns.tolist()
+
+        # filter groupby_attrs to ensure no columns with all nans
+        def _allnan_or_nonan(column):
+            if self.df[column].isnull().all():
+                return False
+            elif self.df[column].isnull().any():
+                raise ValueError(
+                    f'The data in the {column} column should either be all NaN or there should be no NaNs'
+                )
+            else:
+                return True
+
+        groupby_attrs = list(filter(_allnan_or_nonan, groupby_attrs))
+
+        info = {
+            'groupby_attrs': groupby_attrs,
+            'variable_column_name': variable_column_name,
+            'aggregations': aggregations,
+            'agg_columns': agg_columns,
+            'aggregation_dict': aggregation_dict,
+            'path_column_name': path_column_name,
+            'data_format': data_format,
+            'format_column_name': format_column_name,
+        }
+        return info
+
+    def keys(self):
+        keys = list(map(lambda x: self.sep.join(x), self._keys))
+        return keys
+
+    @property
+    def key_template(self):
+        return self.sep.join(self.aggregation_info['groupby_attrs'])
+
+    def __len__(self):
+        return len(self.keys())
+
+    def _get_entries(self):
```

Comment:
This method is called by `self.items()`. So, I went ahead and made sure to load all entries missing from `self._entries`.